### ISSUE TITLE
adapter / command line conflict for MSGF+

### ISSUE THREAD

**Issue by lars20070** [OpenMS/OpenMS#1446]
In order to check the MSGF+ adapter, we did run searches with the adapter and from the command line on Q Exactive data.

(1) idXML straight from MSGFPlusAdapter
(2) MSGF+ from command line (with identical params as the adapter), IDFileConverter, RTAnnotator

For a few sequences, the idXML differ.

![msgfplus_conflict](https://cloud.githubusercontent.com/assets/6149805/7807706/00c78a26-038e-11e5-88f3-00933dda2702.png)

**Comment by lars20070** [comment]
The fault appears to lie with the idXML from  (2). A rank=2 instead of a rank=1 sequence is displayed.

**Comment by timosachsenberg** [comment]
so the sequence annotated as rank=2 is displayed in toppview but there is also one with rank=1?

**Comment by lars20070** [comment]
Yep. Here [1] the data and one example. The mzid contains two PSMs for scan 21732.

LQKERKERKK = Pep15947 (rank=1)
GPSNKKRGRSKK = Pep15948 (rank=2)

Pep15947 is correctly displayed in the idXML from the adapter. Pep15948 is shown in the idXML from the mzid -> idXML conversion with IDFileConverter.

[1] https://www.dropbox.com/sh/s63lknlu3i05pm0/AADFxe58poLbkNyMOZWFaaT0a

**Comment by timosachsenberg** [comment]
My question was if this is a TOPPView issue or a FileConversion issue.

**Comment by timosachsenberg** [comment]
if I search for "GPSNKKRGRSKK" in the command line file I don't get any hit

**Comment by lars20070** [comment]
It's a file conversion problem. In the PeptideIdentification, the two PeptideHits have the wrong order (rank 2, then rank 1). - You cannot find the sequence since idXML stores the modified, but mzid then unmodified (stripped) sequence.

**Comment by timosachsenberg** [comment]
Ok so they have the correct rank but the wrong order - is this correct? Is higher score better set correctly? Then sorting should be sufficient to fix this.

**Comment by lars20070** [comment]
The higher_score_better is incorrect. Should be false, since the scores are e-values (expected number of times to observe by chance). So either correct this, or order by rank instead of score. Rank is given explicitly in mzid.

**Comment by lars20070** [comment]
@mwalzer I am looking for the mapping from score_type to higher_score_better. For example, e-value -> false. Where is this set? (By default, it is true.)

**Comment by timosachsenberg** [comment]
I think the problem ist that there are cases were the E-value value is given in -log 
see:
```
<PeptideIdentification score_type="XTandem" higher_score_better="true" significance_threshold="0" MZ="579.74" RT="1" >
			<PeptideHit score="27.1" sequence="VTAFIPPWVK" charge="2" aa_before="K" aa_after="D" protein_refs="PH_175" >
				<UserParam type="float" name="E-Value" value="14"/>
			</PeptideHit>
```

```
<PeptideIdentification score_type="XTandem" higher_score_better="false" significance_threshold="0" MZ="1393.63" RT="1260" >
			<PeptideHit score="7.5e-10" sequence="SGYSSPGSPGTPGSR" charge="1" aa_before="R" aa_after="S" >
				<UserParam type="float" name="E-Value" value="7.5e-10"/>
				<UserParam type="float" name="xtandem_score" value="47"/>
```

**Comment by lars20070** [comment]
In my example above, the scores are simply 14.004 and 14.049 with the first one beng lower (and better).

**Comment by timosachsenberg** [comment]
ok that is very unexpected: 14.something means that it is -log scaled but in that case higher should be better

**Comment by lars20070** [comment]
MSGF+ problem? (BTW, there should not exist any rank=2 anyway, since we run MSGF+ with -n 1.)

**Comment by mwalzer** [comment]
@lars20070 MzIdentMLHandler is writing Identifications only at the moment. And for writing, you really should have your identifications structures in order, meaning is_higher_better already set.

**Comment by hendrikweisser** [comment]
@mwalzer:
The problem comes from reading the mzid produced by MS-GF+. Specifically, it seems that "PeptideIdentification::higher_score_better_" is not set correctly for the score type that is selected ("SpecEValue").

**Comment by timosachsenberg** [comment]
@hendrikweisser are you sure that this is the problem? Are SpecEValues as high as 14 (without - log scaling!) produced by MSGF+. I would doubt that. I would suspect that the E-Value is -log scaled and the higher this scaled value is, the better.

**Comment by lars20070** [comment]
The lower the better, as I understand [1].

[1] "What about p-Values and E-Values?" page 42 in J Proteome Res, 2008 vol. 7 (1) pp. 40-44
http://pubs.acs.org/doi/abs/10.1021/pr700739d

**Comment by timosachsenberg** [comment]
Yes of course - if not -log transformed

**Comment by oliverschilling** [comment]
Hi, we just discussed this here. First of all, these sequences should stem from a tryptic digest. With 4- 5 missed cleavages, these are really poor PSMs. So 14.xxx ist the actual e-value and it isa really poor e-value. With 14.004 < 14.049 , LQKERKERKK = Pep15947 (rank=1) should be reported, but please keep in mind that these are really poor PSMs, before any FDR filtering.

**Comment by hendrikweisser** [comment]
> Specifically, it seems that "PeptideIdentification::higher_score_better_" is not set correctly for the score type that is selected ("SpecEValue").

I've made a mistake and looked at an output file from the MSGFPlusAdapter. The score type that is selected when converting an MS-GF+ mzid with IDFileConverter is "EValue", not "SpecEValue".

I don't know what this means for scaling/higher-score-better.

**Comment by lars20070** [comment]
@timosachsenberg Here another (more typical) e-value from the same mzid. 14.004 and 14.049 are the exception. So MS-GF+ really reports e-vales here.

value="3.866987E-7" name="MS-GF:SpecEValue"

**Comment by timosachsenberg** [comment]
I see. Thanks for clarifying.

**Comment by lars20070** [comment]
> @lars20070 MzIdentMLHandler is writing Identifications only at the moment. And for writing, you really should have your identifications structures in order, meaning is_higher_better already set.

@mwalzer Looking in the code, the IDFileConverter uses MzIdentMLFile::load which in turn uses MzIdentMLHandler for parsing and reading.

Can we already read mzid and use this mzid -> idXML conversion in IDFileConverter and the MS-GF+ adapter?

**Comment by mwalzer** [comment]
@lars20070 MzIdentMLFile::load is using the MzIdentMLDOMHandler. 
And yes. But such conversions to older id formats are strained by losses in information.

**Comment by timosachsenberg** [comment]
I think it is not a problem to loose information if the one supported by idXML is created - at least then Lars workflow would work. As I understood from our discussion it might be a problem to write out the correct score stored in the mzIdentML as score in the idXML. Is this an issue?

**Comment by lars20070** [comment]
@timosachsenberg No. The score is correct. The score type is correct. Only the higher_score_better flag is incorrect.

The current mzid read implementation is good. So good that @hendrikweisser will replace the old tsv/idXML by this mzid/idXML conversion in MSGFPlusAdapter. (Needed for our Q Star data analysis.) The only thing missing is the correct setting of the higher_score_better flag from its default value true to the correct value false for e-values.

IMO, the higher_score_better flag should be linked to the score type, i.e. score_type = e-value =>  higher_score_better = false. If there is no such mapping in the code, we should add it. https://github.com/OpenMS/OpenMS/issues/1446#issuecomment-106783845

**Comment by timosachsenberg** [comment]
Ok then please make sure that every search engine adapter that writes out the "E-Value" does not perform the scaling. Ótherwise we will break these. Also you should search for "E-Value" in every cpp and h file and double check how it is used.

**Comment by mwalzer** [comment]
This is an general issue with the score typeset in our internal datastructures ... we have loosely defined score types, but nothing a) comprehensive and b) consistent and c) fit for several values.
So I guess you've already seen it, the search engines write a plethora of scores, and there is no ONE score.
Timo and I already spent some time last week to draft how that could look. To preempt that, CV terms can be hugely beneficial! They are hierarchically structured, so we can pull score from the metainfointerface without guessing or compatibility issues, they are typed so no guessing here, and since they are described in the CV, automatic switch and sort according to the inherent score property of isbiggerbetter should be quite feasible. I will reference here when we will put it to discussion soon.

**Comment by timosachsenberg** [comment]
for reference:
That's how it is stored in mzIdentML:
```
            <SpectrumIdentificationResult spectraData_ref="SID_1" spectrumID="controllerType=0 controllerNumber=1 scan=8791" id="SIR_8791">
                <SpectrumIdentificationItem passThreshold="true" rank="1" peptide_ref="Pep25417" calculatedMassToCharge="332.1771545410156" experimentalMassToCharge="331.6700134277344" chargeState="2" id="SII_8791_1">
                    <PeptideEvidenceRef peptideEvidence_ref="PepEv_2026587_25417_477"/>
                    <PeptideEvidenceRef peptideEvidence_ref="PepEv_6909372_25417_200"/>
                    <cvParam accession="MS:1002049" cvRef="PSI-MS" value="-54" name="MS-GF:RawScore"/>
                    <cvParam accession="MS:1002050" cvRef="PSI-MS" value="41" name="MS-GF:DeNovoScore"/>
                    <cvParam accession="MS:1002052" cvRef="PSI-MS" value="0.024811005" name="MS-GF:SpecEValue"/>
                    <cvParam accession="MS:1002053" cvRef="PSI-MS" value="456212.1" name="MS-GF:EValue"/>
                    <userParam value="-1" name="IsotopeError"/>
                    <userParam value="HCD" name="AssumedDissociationMethod"/>
                    <userParam value="0.03560994" name="ExplainedIonCurrentRatio"/>
                    <userParam value="5.286562E-4" name="NTermIonCurrentRatio"/>
                    <userParam value="0.035081282" name="CTermIonCurrentRatio"/>
                    <userParam value="4894363.0" name="MS2IonCurrent"/>
                    <userParam value="3" name="NumMatchedMainIons"/>
                    <userParam value="6.155539" name="MeanErrorAll"/>
                    <userParam value="4.386556" name="StdevErrorAll"/>
                    <userParam value="6.155539" name="MeanErrorTop7"/>
                    <userParam value="4.386556" name="StdevErrorTop7"/>
                    <userParam value="4.819491" name="MeanRelErrorAll"/>
                    <userParam value="5.8228035" name="StdevRelErrorAll"/>
                    <userParam value="4.819491" name="MeanRelErrorTop7"/>
                    <userParam value="5.8228035" name="StdevRelErrorTop7"/>
                </SpectrumIdentificationItem>
                <cvParam accession="MS:1001115" cvRef="PSI-MS" value="8791" name="scan number(s)"/>
            </SpectrumIdentificationResult>
```

**Comment by timosachsenberg** [comment]
And that's how it looks after conversion to idXML - here higher_score_better is set to true (should be false)
```
		<PeptideIdentification score_type="MS-GF:EValue" higher_score_better="true" significance_threshold="0" MZ="657.599975585938" RT="3208.34526" spectrum_reference="controllerType=0 controllerNumber=1 scan=23649" >
			<PeptideHit score="14.706355" sequence="(Dimethyl:2H(4))RVQNQQEAMPSASEDLFNK(Dimethyl:2H(4))NPK(Dimethyl:2H(4))" charge="4" aa_before="K" aa_after="R" protein_refs="PH_11083" >
				<UserParam type="string" name="MS:1002049" value="-12"/>
				<UserParam type="string" name="MS:1002050" value="101"/>
				<UserParam type="string" name="MS:1002052" value="7.896949E-7"/>
				<UserParam type="string" name="MS:1002053" value="14.706355"/>
				<UserParam type="string" name="AssumedDissociationMethod" value="HCD"/>
				<UserParam type="string" name="CTermIonCurrentRatio" value="0.0"/>
				<UserParam type="string" name="ExplainedIonCurrentRatio" value="0.0"/>
				<UserParam type="string" name="IsotopeError" value="0"/>
				<UserParam type="string" name="MS2IonCurrent" value="548694.06"/>
				<UserParam type="string" name="NTermIonCurrentRatio" value="0.0"/>
				<UserParam type="float" name="calcMZ" value="657.605407714844"/>
				<UserParam type="int" name="start" value="426"/>
				<UserParam type="int" name="end" value="447"/>
				<UserParam type="string" name="target_decoy" value="target"/>
			</PeptideHit>
			<UserParam type="string" name="MS:1001115" value="23649"/>
		</PeptideIdentification>
```

**Comment by timosachsenberg** [comment]
as there is this code in the MzIdentMLDomHandler:
```
        else if (specific_score_terms.find(scoreit->first) != specific_score_terms.end() || scoreit->first == "MS:1001143")
        {
          score = scoreit->second.front().getValue().toString().toDouble(); // cast fix needed as DataValue is init with XercesString
          spectrum_identification.setHigherScoreBetter(true);
          spectrum_identification.setScoreType(scoreit->second.front().getName());
          scoretype = true;
        }
```

And the MS-GF:EValue is a child of MS:1001143: 
see:

```
[Term]
id: MS:1002053
name: MS-GF:EValue
def: "MS-GF E-value." [PSI:PI]
xref: value-type:xsd\:double "The allowed value-type for this CV term."
is_a: MS:1001143 ! search engine specific score for PSMs
is_a: MS:1001153 ! search engine specific score
is_a: MS:1002353 ! PSM-level e-value
relationship: has_order MS:1002109 ! lower score better
```

This seems to be the part where the error is made.

**Comment by timosachsenberg** [comment]
@mwalzer shouldn't the dom handler parse the
```
relationship: has_order MS:1002109 ! lower score better
```
in that case?
So this seems to be fixable in my view.

**Comment by mwalzer** [comment]
Timo is definitely right, this has not to be 'hacked', there is a nice and easy solution.

**Comment by hendrikweisser** [comment]
> The current mzid read implementation is good. So good that @hendrikweisser will replace the old tsv/idXML by this mzid/idXML conversion in MSGFPlusAdapter. (Needed for our Q Star data.) The only thing missing is the correct setting of the higher_score_better flag from its default value true to the correct value false for e-values.

About replacing the way idXML is generated in MSGFPlusAdapter: Apart from changes to the mzIdentML implementation in #1505 that should address the score orientation issue, but haven't been merged yet, there is another problem: The direct mzid to idXML conversion can currently require much more time and memory than the (MS-GF+-specific) indirect conversion via TSV - see #1548.

**Comment by mwalzer** [comment]
I think it would be good to keep the option to MSGF+ tsv convert and read, maybe as an advanced option?

**Comment by lars20070** [comment]
That means that both tsv -> idXML and mzid -> idXML need to be maintained and fixed. See tsv -> idXML bug on Q Star data: MS-GF+ finds 3300 PSMs not 3 PSMs [1].

Either way, the MSGFPlusAdapter should reliably report the MS-GF+ search results, and be tested on data from various machines (Q Exactive, Q Star, TripleTOF, ...). I volunteer.

[1] https://github.com/OpenMS/OpenMS/issues/1409#issuecomment-104915565

**Comment by lars20070** [comment]
Resolved in #1446.
